QED code with unit test file

gedeck opened the pull request:
This is essentially identical to PR #1409 from Brian (copied files). 

- Added test suite for unit and regression testing of the QED module
- Refactored based on suggestions in PR #1409 and here
- Added tests for the new descriptors (complexity based on Morgan fingerprint counts)
- Include the doctests from the Descriptors module

Bad news for @bp-kelley, the values change from the Novartis internal implementation due to differences in the number of rotatable bonds calculation. 

Please take a look at the result for the calculation of the number of aromatic rings. In my opinion, the current implementation gets the number wrong for some ring systems, e.g. three annulated ring systems where the central ring is not aromatic. I include a few examples for consideration.

Closes #1442

gedeck committed:
QED code with unit test file

gedeck committed:
Get some debug output for 2.7 and 3.5

gedeck committed:
Actually run the code

greglandrum commented on rdkit/Chem/UnitTestQED.py:
```diff
@@ -0,0 +1,259 @@
+    self.assertEqual([], p)
+    
+
+# 
```
Before merging this, all this commented out stuff should be removed

gedeck replied:
That is for sure. What worries me is that there is an obvious error in this test, `p != []`, but the travis-CI run didn't fail. (OMG - just found my mistake).

As I mentioned, the old doctests fail on Python 3.6. I've made several attempts to install 2.7 in an ubuntu machine and failed miserably. Well, eventually installed rdkit 2017.03 from conda and were able to test. The doctests will also fail on 2.7. Is Novartis still using a different definition of rotatable bonds @bp-kelley ? If yes, that's the reason. 

I'm not really concerned about the impact of having the number of rotatable bonds by a few. The differences in the qed values was 0.02. That's negligible for a qualitative descriptor like this. To make sure that we use a consistent approach, we better replace the `Lipinski.NumRotatableBonds(mol)` with a more suitable function.  Would directly using `rdMolDescriptors.CalcNumRotatableBonds(x)` be sufficient?

greglandrum commented on rdkit/Chem/Descriptors.py:
```diff
@@ -181,6 +181,21 @@ def MinAbsPartialCharge(mol, force=False):
+from rdkit.Chem.QED import qed
+
+
+def _FingerprintDensity(mol, func, *args, **kwargs):
```
This differences between the bit vector and count vector form of this function are bothering me.
Here's the asymptotic behavior for the the two different densities for linear carbon chains with between 1 and 100 atoms:
![image](https://cloud.githubusercontent.com/assets/540511/26771893/6c9e8c00-49c2-11e7-9d7a-6612996a888f.png)
and
![image](https://cloud.githubusercontent.com/assets/540511/26771904/779bc7da-49c2-11e7-901c-d97735cbf4b3.png)

I'm not sure which is "more correct", but the fact that they are so different is disturbing.

gedeck replied:
I assume this is for the morgan-2 fingerprints. The count vector form must go to 3, as there are 3 fragments created around each atom. For `C` it is only one (1/1), for `CC` it's two (4/2 theoretically), for `CCC` it's (3+2+3/3 theoretically). In reality it's a bit more complicated, as we take the bonding to the next partner into account and, I think, the terminal atoms are treated somewhat differently; (2 + 1)/2 for CC, (2 + 1 + 1 + 2)/3 for CCC) . For 'C'*n and n sufficiently large, the counts are: (n-2, n-4, n-6, 5*2), so we get (3*n - 2) / n = 3 - 2/n.

That makes the count vector relatively uninteresting. Even more complex structures, e.g. `'CNCCNNCCNC'*10` have the same asymptotic behaviour. The counts vector based descriptor is only a measure of how many terminal groups and what percentage of the structure they make. 

The bit vector form is more interesting; it is a measure of structural diversity. Here, the asymptotic for `'C'*n` is 8/n so it will approach 0 with increasing chain length.

greglandrum replied:
Yeah, I agree with you. I think the count-based fingerprints should be using `fp.GetNonzeroElements()` (which returns the number of set "bits") instead of `GetTotalVal()`

greglandrum reviewed (changes requested):
I'm commenting on @bp-kelley's original code as well as @gedeck's suggested improvements

gedeck committed:
Added initial unit and regression tests

gedeck commented:
I've added unit and regression tests for the QED module. This will allow refactoring the code as discussed previously. 

An additional issue, I've identified is that the code gives different results if the molecule has hydrogens added or not (changes rotatable bonds and number of structural alerts). I don't think this should happen.

gedeck committed:
Refactored following the suggestions from PR1409

greglandrum commented:
This looks good to me. @bp-kelley : what do you think?

bp-kelley commented:
In principle I like the changes, we may think of bumping the version since it is now different and conflicts with our internal version.

gedeck committed:
Add descriptor versions.

gedeck commented:
Add version information for the QED descriptors (1.1.0) and the FpMorganDensity descriptors.

greglandrum commented on rdkit/Chem/UnitTestDescriptors.py:
```diff
@@ -113,11 +119,31 @@ def testMQN(self):
+    m = Chem.MolFromSmiles('CC')
+    self.assertEqual(Descriptors.FpDensityMorgan2(m), 1)
+    m = Chem.MolFromSmiles('CCC')
+    self.assertEqual(Descriptors.FpDensityMorgan2(m), 4.0 / 3)
```
using `assertEqual` with floats makes me nervous. Sure, it shouldn't ever cause a problem, but who knows...
Would you mind switch to `assertAlmostEqual`?

bp-kelley replied:
I'm happy with the version changes.  I think we might be able to expect floating point identity here, but I've been surprised before.

gedeck committed:
Various changes

- Replace assertEqual with assertAlmostEqual if appropriate
- Add version for the fingerprint density descriptors

gedeck commented:
Good catch with the floating point tests - changed them.
